# Yaqeen issa 1201576

In [ ]:
import pandas as pd
import numpy as np
import random
from itertools import combinations
from collections import Counter

In [ ]:
#read data fream
df=pd.read_csv("online_retail_II.csv")
df.sort_values(by="Quantity")

Invoice StockCode                          Description  Quantity  \
587085  C541433     23166       MEDIUM CERAMIC TOP STORAGE JAR    -74215   
303996   519017     22759                                  NaN     -9600   
529729  C536757     84347  ROTATING SILVER ANGELS T-LIGHT HLDR     -9360   
507225  C536757     84347  ROTATING SILVER ANGELS T-LIGHT HLDR     -9360   
156488   504311     22197                                  NaN     -9200   
...         ...       ...                                  ...       ...   
127169   501534     21085            SET/6 WOODLAND PAPER CUPS     12744   
127168   501534     21091          SET/6 WOODLAND PAPER PLATES     12960   
127166   501534     21099          SET/6 STRAWBERRY PAPER CUPS     12960   
90857    497946     37410   BLACK AND WHITE PAISLEY FLOWER MUG     19152   
587080   541431     23166       MEDIUM CERAMIC TOP STORAGE JAR     74215   

                InvoiceDate  Price  Customer ID         Country  
587085  2011-01-18 10:17:00   1.04      12346.0  United Kingdom  
303996  2010-08-13 09:14:00   0.00          NaN  United Kingdom  
529729  2010-12-02 14:23:00   0.03      15838.0  United Kingdom  
507225  2010-12-02 14:23:00   0.03      15838.0  United Kingdom  
156488  2010-04-12 14:39:00   0.00          NaN  United Kingdom  
...                     ...    ...          ...             ...  
127169  2010-03-17 13:09:00   0.10      13902.0         Denmark  
127168  2010-03-17 13:09:00   0.10      13902.0         Denmark  
127166  2010-03-17 13:09:00   0.10      13902.0         Denmark  
90857   2010-02-15 11:57:00   0.10      13902.0         Denmark  
587080  2011-01-18 10:01:00   1.04      12346.0  United Kingdom  

[719774 rows x 8 columns]

Randomly select 200k row from the dataset using your ID as seed.

In [ ]:
# Select 200000 random rows
np.random.seed(1201576)
sample_200K=df.sample(n=200000,replace=False)

In [ ]:
df.head(1)

Invoice StockCode                          Description  Quantity  \
0  489434     85048  15CM CHRISTMAS GLASS BALL 20 LIGHTS        12   

           InvoiceDate  Price  Customer ID         Country  
0  2009-12-01 07:45:00   6.95      13085.0  United Kingdom

1.   Determine the total sales per invoice and identify the top 5 invoices.

In [ ]:
sample_200K['TotalSales'] = sample_200K['Quantity'] * sample_200K['Price']

In [ ]:
top_5_invoices = sample_200K.groupby('Invoice')['TotalSales'].sum().sort_values(ascending=False).head(5)
top_5_invoices_df=top_5_invoices.reset_index()
print(top_5_invoices_df)

  Invoice  TotalSales
0  533027    18592.45
1  515944    12366.00
2  493819    12347.66
3  502263    10953.50
4  524181     7674.96


The top invoice number is 541431 and the total quantity for it is 74215

#Edit number





2.Compute the total sales per customer and list the top 5 customers.


In [ ]:

top_5_customer = sample_200K.groupby("Customer ID")['TotalSales'].sum().sort_values(ascending=False).head(5)

top_5_customer_df = top_5_customer.reset_index()

print('Top 5 Customers by Total Sales:')
print(top_5_customer_df)

Top 5 Customers by Total Sales:
   Customer ID  TotalSales
0      14646.0    98567.53
1      18102.0    95642.60
2      14156.0    59261.31
3      14911.0    47807.67
4      13694.0    39058.19


This output indicates the top 5 customers by total sales, with each line showing a customer ID followed by their corresponding total sales amount.

Identify the top 5 selling items during this period.


In [ ]:
top_5_items=sample_200K.groupby("StockCode")['Quantity'].sum().sort_values(ascending=False).head(5)
top_5_items_df = top_5_items.to_frame().reset_index()
print("Top 5 Items by Total Sales:")
print(top_5_items_df)

Top 5 Items by Total Sales:
  StockCode  Quantity
0     17003     21079
1     21212     20821
2    85123A     20767
3     84879     20203
4     84077     19340


 output shows the top 5 items by total sales, where each row represents an item identified by its 'StockCode' and its corresponding total sales quantity ('Quantity').

Determine the most frequently purchased item pairs by customers and display the top 20 pairs.

In [ ]:
grouped = sample_200K.groupby('Invoice')['Description'].apply(list)
pairs = Counter()
for items in grouped:
    for pair in combinations(set(items), 2):
        pairs[pair] += 1

top_20_pairs = pairs.most_common(20)
top_20_pairs_df = pd.DataFrame(top_20_pairs, columns=['Pair', 'Count'])

top_20_pairs_df


Pair  Count
0      (HEART OF WICKER LARGE, HEART OF WICKER SMALL)     70
1   (PACK OF 72 RETRO SPOT CAKE CASES, 60 TEATIME ...     61
2   (STRAWBERRY CERAMIC TRINKET BOX, SWEETHEART CE...     59
3   (WHITE HANGING HEART T-LIGHT HOLDER, RED HANGI...     53
4   (WOODEN PICTURE FRAME WHITE FINISH, WOODEN FRA...     52
5   (WHITE HANGING HEART T-LIGHT HOLDER, WOODEN PI...     51
6   (JUMBO STORAGE BAG SUKI, JUMBO  BAG BAROQUE BL...     51
7   (JUMBO BAG RED RETROSPOT, JUMBO  BAG BAROQUE B...     50
8   (PACK OF 60 PINK PAISLEY CAKE CASES, 60 TEATIM...     49
9   (WHITE HANGING HEART T-LIGHT HOLDER, STRAWBERR...     49
10     (JUMBO BAG STRAWBERRY, JUMBO STORAGE BAG SUKI)     46
11  (WOOD 2 DRAWER CABINET WHITE FINISH, WOOD S/3 ...     46
12  (JUMBO BAG RED RETROSPOT, JUMBO STORAGE BAG SUKI)     45
13  (HOT WATER BOTTLE TEA AND SYMPATHY, CHOCOLATE ...     44
14  (HOME BUILDING BLOCK WORD, LOVE BUILDING BLOCK...     44
15  (JUMBO BAG RED RETROSPOT, JUMBO SHOPPER VINTAG...     44
16  (PACK OF 60 DINOSAUR CAKE CASES, 60 TEATIME FA...     43
17  (PLASTERS IN TIN CIRCUS PARADE , PLASTERS IN T...     43
18  (PINK BLUE FELT CRAFT TRINKET BOX, PINK CREAM ...     42
19  (PACK OF 72 SKULL CAKE CASES, 60 TEATIME FAIRY...     42

 code calculates the top 20 pairs of items based on their occurrence in invoices. Each pair consists of two items

The retail store aims to introduce a new promotion targeting items with low sales volume. This promotion entails offering a discounted price for a pair of two different items, with one item being among the top 10 best sellers and the other from the bottom 10 sellers. Customers who purchase these two items together will qualify for the special pricing. Your task is to recommend 10 pairs of items to kickstart this campaign.

In [ ]:
top_10_items = df.groupby('Description')['Quantity'].sum().nlargest(10).index.tolist()
bottom_10_items = df.groupby('Description')['Quantity'].sum().nsmallest(10).index.tolist()
promotion_pairs = [(top, bottom) for top in top_10_items for bottom in bottom_10_items][:10]

top_10_items_df = pd.DataFrame(promotion_pairs, columns=['Description','Description2'])
top_10_items_df


Description            Description2
0  WORLD WAR 2 GLIDERS ASSTD DESIGNS                       ?
1  WORLD WAR 2 GLIDERS ASSTD DESIGNS              given away
2  WORLD WAR 2 GLIDERS ASSTD DESIGNS                 missing
3  WORLD WAR 2 GLIDERS ASSTD DESIGNS              ebay sales
4  WORLD WAR 2 GLIDERS ASSTD DESIGNS              Given away
5  WORLD WAR 2 GLIDERS ASSTD DESIGNS     Zebra invcing error
6  WORLD WAR 2 GLIDERS ASSTD DESIGNS  Ebay sales by the box.
7  WORLD WAR 2 GLIDERS ASSTD DESIGNS              throw away
8  WORLD WAR 2 GLIDERS ASSTD DESIGNS                 damages
9  WORLD WAR 2 GLIDERS ASSTD DESIGNS             wet damages

the output show pairs of items from the top 10 and bottom 10 items, and to recommend 10 pairs of items to kickstart the campaign